# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/JSContext.h
# Commit: 982a80b202c3
# Full Hash: 982a80b202c370876dafc72ac31be7a12ebeae58
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2021-01-08 03:45:03
# Regressor Bug: 1681533
# File Overlap Count: 2
# Description:
#   Bug 1681533 - Only collect the nursery when necessary in major GC r=sfink
#   
#   This rearranges a few things so that we can run a minor GC inside a major GC
#   slice, so we don't have to be conservative about collecting the nursery at the
#   start of a slice.
# ==============================================================================

diff -r 83f3726c41f9 -r 982a80b202c3 js/src/vm/JSContext.h
--- a/js/src/vm/JSContext.h	Thu Jan 07 15:57:41 2021 +0000
+++ b/js/src/vm/JSContext.h	Mon Jan 04 17:01:50 2021 +0000
@@ -1186,22 +1186,18 @@
 
 namespace gc {
 
-// Set/unset the performing GC flag for the current thread.
+// Set/restore the performing GC flag for the current thread.
 class MOZ_RAII AutoSetThreadIsPerformingGC {
   JSContext* cx;
+  bool prev;
 
  public:
-  AutoSetThreadIsPerformingGC() : cx(TlsContext.get()) {
-    JSFreeOp* fop = cx->defaultFreeOp();
-    MOZ_ASSERT(!fop->isCollecting());
-    fop->isCollecting_ = true;
+  AutoSetThreadIsPerformingGC()
+      : cx(TlsContext.get()), prev(cx->defaultFreeOp()->isCollecting_) {
+    cx->defaultFreeOp()->isCollecting_ = true;
   }
 
-  ~AutoSetThreadIsPerformingGC() {
-    JSFreeOp* fop = cx->defaultFreeOp();
-    MOZ_ASSERT(fop->isCollecting());
-    fop->isCollecting_ = false;
-  }
+  ~AutoSetThreadIsPerformingGC() { cx->defaultFreeOp()->isCollecting_ = prev; }
 };
 
 struct MOZ_RAII AutoSetThreadGCUse {
